### PR TITLE
Refactored authenticating user in SOAP [XCEP]endpoint.

### PR DESCRIPTION
**Pull request by hasuniea**

### CODE REVIEW

File: modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/wstep/impl/CertificateEnrollmentServiceImpl.java [hunk 1]
```diff
@@ -118,25 +122,28 @@ public void requestSecurityToken(String tokenType, String requestType,
                 headerBinarySecurityToken = element.getFirstChild().getNextSibling().getFirstChild().getTextContent();
             }
         }
-        List<ConfigurationEntry> tenantConfigurations;
+        List<ConfigurationEntry> tenantConfigurations = null;
         try {
             if (getTenantConfigurationData() != null) {
                 tenantConfigurations = getTenantConfigurationData();
+                for (ConfigurationEntry configurationEntry : tenantConfigurations) {
+                    if (configurationEntry.getName().equals(PluginConstants.TenantConfigProperties.
```
Comment (harshanL) [inline modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/wstep/impl/CertificateEnrollmentServiceImpl.java]: Fix the equals method.

Comment (hasuniea) [inline modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/wstep/impl/CertificateEnrollmentServiceImpl.java]: fixed the issue

File: modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/wstep/impl/CertificateEnrollmentServiceImpl.java [hunk 2]
```diff
@@ -303,14 +311,35 @@ public String prepareWapProvisioningXML(
...
-            //Generic exception is caught here as there is no need of taking different actions for
-            //different exceptions.
-        } catch (Exception e) {
-            String msg = "Problem occurred with wap-provisioning.xml file.";
+        } catch (ParserConfigurationException e) {
+            String msg = "Problem occurred in parsing wap-provisioning.xml file.";
+            log.error(msg, e);
+            throw new WAPProvisioningException(msg, e);
+        } catch (DeviceManagementException e) {
+            String msg = "Error occurred in while getting CA and Root certificates.";
```
Comment (kasundananjaya) [inline modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/wstep/impl/CertificateEnrollmentServiceImpl.java]: Make error messages more meaningful. (Check grammar if possible)

Comment (hasuniea) [inline modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/wstep/impl/CertificateEnrollmentServiceImpl.java]: fixed issue from https://github.com/wso2/product-mdm/pull/156

File: modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/common/util/BSTValidator.java [hunk 3]
```diff
@@ -58,12 +83,13 @@ public Credential validate(Credential credential, RequestData requestData) throw
...
-            //Generic exception is caught here as there is no need of taking different actions for
-            //different exceptions.
-        } catch (Exception e) {
+        } catch (AuthenticationException e) {
             String msg = "Failure occurred in the BST validator.";
             log.error(msg, e);
             throw new WSSecurityException(msg, e);
+        } catch (WindowsDeviceEnrolmentException e) {
+            String msg = "Error occurred in validating request token.";
+            log.error(msg);
```
Comment (prabathabey) [inline modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/common/util/BSTValidator.java]: Avoid "log and throw" as much as possible, unless specifically required.

File: modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/common/util/BSTValidator.java [hunk 4]
```diff
@@ -75,15 +101,63 @@ public Credential validate(Credential credential, RequestData requestData) throw
...
+                String username = oAuth2TokenValidationResponseDTO.getAuthorizedUser();
+                authenticationInfo.setUsername(username);
+                authenticationInfo.setTenantDomain(MultitenantUtils.getTenantDomain(username));
+                authenticationInfo.setTenantId(WindowsAPIUtils.getTenantIdOFUser(username));
+            } else {
+                authenticationInfo.setMessage(oAuth2TokenValidationResponseDTO.getErrorMsg());
+            }
+        } catch (DeviceManagementException e) {
+            String msg = "Authentication failure due to invalid binary security token.";
+            log.error(msg);
```
Comment (prabathabey) [inline modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/common/util/BSTValidator.java]: Avoid "log and throw" as much as possible, unless specifically required.

File: modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/common/util/WindowsAPIUtils.java [hunk 5]
```diff
@@ -166,4 +170,33 @@ public static TenantConfiguration getTenantConfiguration() throws DeviceManageme
...
+        return tenantId;
+    }
+
+    public static OAuth2TokenValidationService getOAuth2TokenValidationService() {
+        PrivilegedCarbonContext ctx = PrivilegedCarbonContext.getThreadLocalCarbonContext();
+        OAuth2TokenValidationService oAuth2TokenValidationService =
+                (OAuth2TokenValidationService) ctx.getOSGiService(OAuth2TokenValidationService.class, null);
+        if (oAuth2TokenValidationService == null) {
+            String msg = "OAuth2TokenValidation service has not initialized.";
+            log.error(msg);
```
Comment (prabathabey) [inline modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/common/util/WindowsAPIUtils.java]: Avoid "log and throw" as much as possible, unless specifically required.

File: modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/common/util/WindowsAPIUtils.java [hunk 6]
```diff
@@ -166,4 +170,33 @@ public static TenantConfiguration getTenantConfiguration() throws DeviceManageme
...
+        int tenantId = 0;
+        String domainName = MultitenantUtils.getTenantDomain(username);
+        if (domainName != null) {
+            try {
+                TenantManager tenantManager = IdentityTenantUtil.getRealmService().getTenantManager();
+                tenantId = tenantManager.getTenantId(domainName);
+            } catch (UserStoreException e) {
+                String errorMsg = "Error when getting the tenant id from the tenant domain : " +
+                        domainName;
+                log.error(errorMsg, e);
```
Comment (prabathabey) [inline modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/common/util/WindowsAPIUtils.java]: Avoid "log and throw" as much as possible, unless specifically required.

File: modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/wstep/impl/CertificateEnrollmentServiceImpl.java [hunk 7]
```diff
@@ -118,25 +122,28 @@ public void requestSecurityToken(String tokenType, String requestType,
...
+                            NOTIFIER_FREQUENCY)) {
+                        pollingFrequency = configurationEntry.getValue().toString();
+                    } else {
+                        pollingFrequency = PluginConstants.TenantConfigProperties.DEFAULT_FREQUENCY;
+                    }
+                }
             } else {
-                String msg = "Tenant configurations are not initialized.";
+                pollingFrequency = PluginConstants.TenantConfigProperties.DEFAULT_FREQUENCY;
+                String msg = "Tenant configurations are not initialized yet.";
```
Comment (prabathabey) [inline modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/wstep/impl/CertificateEnrollmentServiceImpl.java]: Avoid "log and throw" as much as possible, unless specifically required.

File: modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/common/util/BSTValidator.java (hunk 4 above)
Comment (hasuniea) [inline modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/common/util/BSTValidator.java]: fixed it from https://github.com/wso2/product-mdm/pull/156

File: modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/common/util/WindowsAPIUtils.java (hunk 5 above)
Comment (hasuniea) [inline modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/common/util/WindowsAPIUtils.java]: fixed it from #156

File: modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/common/util/WindowsAPIUtils.java (hunk 6 above)
Comment (hasuniea) [inline modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/common/util/WindowsAPIUtils.java]: fixed it from #156

File: modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/common/util/BSTValidator.java (hunk 3 above)
Comment (hasuniea) [inline modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/common/util/BSTValidator.java]: fixed it from #156

File: modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/common/PluginConstants.java [hunk 8]
```diff
@@ -272,6 +273,7 @@ private TenantConfigProperties() {
         public static final String LANGUAGE_US = "en_US";
         public static final String CONTENT_TYPE_TEXT = "text";
         public static final String NOTIFIER_FREQUENCY = "notifierFrequency";
+        public static final String DEFAULT_FREQUENCY = "8";
```
Comment (DilanUA) [inline modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/common/PluginConstants.java]: unnecessary new line

File: modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/common/util/BSTValidator.java [hunk 9]
```diff
@@ -75,15 +101,63 @@ public Credential validate(Credential credential, RequestData requestData) throw
      * @return - Authentication status
      * @throws AuthenticationException
      */
+
     public boolean authenticate(String binarySecurityToken) throws
             AuthenticationException {
```
Comment (DilanUA) [inline modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/common/util/BSTValidator.java]: unnecessary new line

File: modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/wstep/impl/CertificateEnrollmentServiceImpl.java [hunk 10]
```diff
@@ -201,10 +211,8 @@ private String convertDocumentToString(Document document) throws TransformerExce
     public String prepareWapProvisioningXML(
```
Comment (DilanUA) [inline modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/wstep/impl/CertificateEnrollmentServiceImpl.java]: this method could be private

File: modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/wstep/impl/CertificateEnrollmentServiceImpl.java [hunk 11]
```diff
@@ -303,14 +311,35 @@ public String prepareWapProvisioningXML(
...
             }
             wapProvisioningString = convertDocumentToString(document);
-
-            //Generic exception is caught here as there is no need of taking different actions for
-            //different exceptions.
-        } catch (Exception e) {
-            String msg = "Problem occurred with wap-provisioning.xml file.";
+        } catch (ParserConfigurationException e) {
+            String msg = "Problem occurred in parsing wap-provisioning.xml file.";
+            log.error(msg, e);
```
Comment (DilanUA) [inline modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/wstep/impl/CertificateEnrollmentServiceImpl.java]: once this method becomes private, log messages become redundant

File: modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/wstep/impl/CertificateEnrollmentServiceImpl.java [hunk 12]
```diff
@@ -303,14 +311,35 @@ public String prepareWapProvisioningXML(
...
+            throw new WAPProvisioningException(msg, e);
+        } catch (DeviceManagementException e) {
+            String msg = "Error occurred in while getting CA and Root certificates.";
+            log.error(msg, e);
+            throw new WindowsDeviceEnrolmentException(msg, e);
+        } catch (NoSuchAlgorithmException | CertificateEncodingException e) {
+            String msg = "Error occurred in while encoding certificates.";
+            log.error(msg, e);
```
Comment (DilanUA) [inline modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/wstep/impl/CertificateEnrollmentServiceImpl.java]: This method could be private. once this method becomes private, log messages become redundant

File: modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/wstep/impl/CertificateEnrollmentServiceImpl.java [hunk 13]
```diff
@@ -303,14 +311,35 @@ public String prepareWapProvisioningXML(
...
+            throw new WindowsDeviceEnrolmentException(msg, e);
+        } catch (UnsupportedEncodingException e) {
+            String msg = "Error occurred in while encoding wap-provisioning file.";
+            log.error(msg, e);
```
Comment (DilanUA) [inline modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/wstep/impl/CertificateEnrollmentServiceImpl.java]: This method could be private. once this method becomes private, log messages become redundant

File: modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/wstep/impl/CertificateEnrollmentServiceImpl.java [hunk 14]
```diff
@@ -303,14 +311,35 @@ public String prepareWapProvisioningXML(
...
+            throw new WindowsDeviceEnrolmentException(msg, e);
+        } catch (SAXException e) {
+            String msg = "Error occurred in while parsing wap-provisioning.xml file.";
+            log.error(msg, e);
```
Comment (DilanUA) [inline modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/wstep/impl/CertificateEnrollmentServiceImpl.java]: This method could be private. once this method becomes private, log messages become redundant

File: modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/wstep/impl/CertificateEnrollmentServiceImpl.java [hunk 15]
```diff
@@ -303,14 +311,35 @@ public String prepareWapProvisioningXML(
...
+            throw new WAPProvisioningException(msg, e);
+        } catch (TransformerException e) {
+            String msg = "Error occurred in while transforming wap-provisioning.xml file.";
+            log.error(msg, e);
```
Comment (DilanUA) [inline modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/wstep/impl/CertificateEnrollmentServiceImpl.java]: This method could be private. once this method becomes private, log messages become redundant

File: modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/wstep/impl/CertificateEnrollmentServiceImpl.java [hunk 16]
```diff
@@ -303,14 +311,35 @@ public String prepareWapProvisioningXML(
...
+            throw new WAPProvisioningException(msg, e);
+        } catch (IOException e) {
+            String msg = "Error occurred in while getting wap-provisioning.xml file.";
             log.error(msg, e);
```
Comment (DilanUA) [inline modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/wstep/impl/CertificateEnrollmentServiceImpl.java]: This method could be private. once this method becomes private, log messages become redundant

File: modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/xcep/impl/CertificateEnrollmentPolicyServiceImpl.java [hunk 17]
```diff
@@ -43,63 +35,67 @@
...
+        oidElement.setGroup(PluginConstants.CertificateEnrolmentPolicy.OID_GROUP);
+        oidElement.setOIDReferenceID(PluginConstants.CertificateEnrolmentPolicy.OID_REFERENCE_ID);
+        oidElement.setDefaultName(PluginConstants.CertificateEnrolmentPolicy.OID_DEFAULT_NAME);
+
+        oidCollectionElement.getOID().add(oidElement);
+        caCollection.value = caCollectionElement;
+        oidCollection.value = oidCollectionElement;
+
+        PrivilegedCarbonContext.endTenantFlow();
+
```
Comment (DilanUA) [inline modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/xcep/impl/CertificateEnrollmentPolicyServiceImpl.java]: unnecessary new line

File: modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/common/util/BSTValidator.java (hunk 9 above)
Comment (hasuniea) [inline modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/common/util/BSTValidator.java]: already fixed.

File: modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/common/PluginConstants.java (hunk 8 above)
Comment (hasuniea) [inline modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/common/PluginConstants.java]: already fixed

File: modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/wstep/impl/CertificateEnrollmentServiceImpl.java (hunk 10 above)
Comment (hasuniea) [inline modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/wstep/impl/CertificateEnrollmentServiceImpl.java]: fixed it from https://github.com/wso2/product-mdm/pull/156

File: modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/wstep/impl/CertificateEnrollmentServiceImpl.java (hunk 11 above)
Comment (hasuniea) [inline modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/wstep/impl/CertificateEnrollmentServiceImpl.java]: fixed issue from https://github.com/wso2/product-mdm/pull/156

File: modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/wstep/impl/CertificateEnrollmentServiceImpl.java (hunk 12 above)
Comment (hasuniea) [inline modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/wstep/impl/CertificateEnrollmentServiceImpl.java]: fixed issue from https://github.com/wso2/product-mdm/pull/156

File: modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/wstep/impl/CertificateEnrollmentServiceImpl.java (hunk 13 above)
Comment (hasuniea) [inline modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/wstep/impl/CertificateEnrollmentServiceImpl.java]: fixed issue from https://github.com/wso2/product-mdm/pull/156

File: modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/wstep/impl/CertificateEnrollmentServiceImpl.java (hunk 14 above)
Comment (hasuniea) [inline modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/wstep/impl/CertificateEnrollmentServiceImpl.java]: fixed issue from https://github.com/wso2/product-mdm/pull/156

File: modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/wstep/impl/CertificateEnrollmentServiceImpl.java (hunk 15 above)
Comment (hasuniea) [inline modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/wstep/impl/CertificateEnrollmentServiceImpl.java]: fixed issue from https://github.com/wso2/product-mdm/pull/156

File: modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/wstep/impl/CertificateEnrollmentServiceImpl.java (hunk 16 above)
Comment (hasuniea) [inline modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/wstep/impl/CertificateEnrollmentServiceImpl.java]: fixed issue from https://github.com/wso2/product-mdm/pull/156

File: modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/xcep/impl/CertificateEnrollmentPolicyServiceImpl.java (hunk 17 above)
Comment (hasuniea) [inline modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/xcep/impl/CertificateEnrollmentPolicyServiceImpl.java]: fixed issue from https://github.com/wso2/product-mdm/pull/156

**Pull request merged**